include/mbgl/renderer/layer_group.hpp

```diff
@@ -154,15 +203,78 @@ class LayerGroup : public LayerGroupBase {
     std::vector<gfx::UniqueDrawable> removeDrawables(mbgl::RenderPass);
     void addDrawable(gfx::UniqueDrawable&&);
 
-    std::size_t visitDrawables(const std::function<void(gfx::Drawable&)>&&) override;
-    std::size_t visitDrawables(const std::function<void(const gfx::Drawable&)>&&) const override;
-    std::size_t removeDrawablesIf(const std::function<bool(gfx::Drawable&)>&&) override;
+    template <typename Func /* void(gfx::Drawable&) */>
+    std::size_t visitDrawables(Func f) {
+        for (const auto& item : drawables) {
+            if (item) {
+                f(*item);
+            }
+        }
+        return drawables.size();
+    }
+
+    template <typename Func /* bool(gfx::Drawable&) */>
+    std::size_t removeDrawablesIf(Func f) {
+        decltype(drawables) newSet;
+        const auto oldSize = drawables.size();
+        while (!drawables.empty()) {
+            // set members are immutable, since changes could affect its position, so extract each item
+            gfx::UniqueDrawable drawable = std::move(drawables.extract(drawables.begin()).value());
+            if (!f(*drawable)) {
+                // Not removed, keep it, but in a new set so that if the key value
+                // has increased, we don't see it again during this iteration.
+                newSet.emplace_hint(newSet.end(), std::move(drawable));
+            }
+        }
+        std::swap(drawables, newSet);
+        return (oldSize - drawables.size());
+    }
 
     std::size_t clearDrawables() override;
 
 protected:
-    struct Impl;
-    std::unique_ptr<Impl> impl;
+    using DrawableCollection = std::set<gfx::UniqueDrawable, gfx::DrawableLessByPriority>;
+    DrawableCollection drawables;
 };
 
+template <typename Func /* void(gfx::Drawable&) */>
+void visitLayerGroupDrawables(mbgl::LayerGroupBase& layerGroup, Func dg) {
+    switch (layerGroup.getType()) {
```

Comment:
I'm not super happy with this, but I can't specialize a template member function that also accepts a lambda without using `std::function` to erase the lambda type.

Reply:
maybe something like this would work?

```
        if constexpr (!std::is_invocable<Func, gfx::Drawable&>::value) {
               static_assert(...)
        }
```

Reply:
The problem stems from the polymorphism of LayerGroupBase - a template virtual function isn't possible, so we need to know what kind of layer group we're dealing with to then invoke the right instance's visit template.

Reply:
Are there more lightweight `std::function` alternatives?

Reply:
I don't think so. There are two aspects to overhead with `std::function`, allocation and indirection. We probably weren't incurring allocation overhead in any of these changes, but we were paying the indirection cost.

`std::function`'s type erasure works by using virtual dispatch. The compiler can't inline these virtual function calls as the functions provided could vary and *possibly* carry an allocated closure.

Reply:
This solution is pretty easy to understand. There may be a fancier solution but looks OK to me.